{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b29d24b",
   "metadata": {},
   "source": [
    "# GUI BIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b38b68d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tkinter as tk\n",
    "import tkinter as tk\n",
    "from tkinter import filedialog\n",
    "import customtkinter as ctk\n",
    "from Bio import SeqIO\n",
    "import warnings\n",
    "\n",
    "# Suppress all warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "ctk.set_appearance_mode(\"Dark\")\n",
    "\n",
    "class BioinformaticsApp:\n",
    "    def __init__(self, master):\n",
    "        self.master = master\n",
    "        self.master.title(\"Bioinformatics GUI\")\n",
    "\n",
    "        # Calculate center position for the window\n",
    "        screen_width = self.master.winfo_screenwidth()\n",
    "        screen_height = self.master.winfo_screenheight()\n",
    "        window_width = 857  # Adjusted dimensions\n",
    "        window_height = 857  # Adjusted dimensions\n",
    "        x = (screen_width - window_width) // 2\n",
    "        y = (screen_height - window_height) // 2\n",
    "        self.master.geometry(f\"{window_width}x{window_height}+{x}+{y}\")\n",
    "\n",
    "        # Make the window non-resizable\n",
    "        self.master.resizable(False, False)\n",
    "\n",
    "        self.dna_sequence = \"\"  # Variable to store DNA sequence\n",
    "\n",
    "        self.create_widgets()\n",
    "\n",
    "    def create_widgets(self):\n",
    "        # Frame for file upload\n",
    "        upload_frame = ctk.CTkFrame(self.master)\n",
    "        upload_frame.pack(pady=10)\n",
    "\n",
    "        file_name = tk.StringVar()\n",
    "        file_name.set(\"No file selected\")\n",
    "\n",
    "        def open_file():\n",
    "            nonlocal file_name\n",
    "            file_path = filedialog.askopenfilename(filetypes=[(\"FASTA files\", \"*.fasta\")])\n",
    "            if file_path:\n",
    "                file_name.set(file_path.split(\"/\")[-1])\n",
    "\n",
    "                # Read DNA sequence from the selected file\n",
    "                try:\n",
    "                    self.dna_sequence = ''\n",
    "                    with open(file_path, \"r\") as f:\n",
    "                        for line in f:\n",
    "                            if not line.startswith('>'):\n",
    "                                self.dna_sequence += line.strip()\n",
    "\n",
    "                    # Print the loaded sequence\n",
    "                    print(\"Loaded DNA Sequence:\", self.dna_sequence)\n",
    "                except FileNotFoundError:\n",
    "                    print(f\"File not found: {file_path}\")\n",
    "\n",
    "        upload_button = ctk.CTkButton(upload_frame, text=\"Upload DNA file\", command=open_file)\n",
    "        upload_button.pack(pady=5)\n",
    "\n",
    "        file_label = ctk.CTkLabel(upload_frame, textvariable=file_name)\n",
    "        file_label.pack(pady=5)\n",
    "\n",
    "        # Frame for algorithm checklist\n",
    "        algo_frame = ctk.CTkFrame(self.master)\n",
    "        algo_frame.pack(pady=10)\n",
    "\n",
    "        algo_label = ctk.CTkLabel(algo_frame, text=\"Select the algorithm to apply:\")\n",
    "        algo_label.pack()\n",
    "\n",
    "        algos = [\"DNA reverse complementary\", \"DNA translation to protein\", \"Exact matching\", \"Boyer moore\", \"K-mer indexing\", \"K-M-P algorithm\", \"Suffix array\", \"Approximate matching\"]\n",
    "\n",
    "        selected_algo = tk.StringVar(value=algos[0])  # Initialize with the first algorithm\n",
    "                                                    # start choice\n",
    "        for algo in algos:\n",
    "            radio_button = tk.Radiobutton(algo_frame, text=algo, variable=selected_algo, value=algo,bg=\"#303030\",\n",
    "                                         fg=\"white\", selectcolor=\"#303030\",activebackground=\"#303030\")\n",
    "            radio_button.pack(anchor=tk.W)\n",
    "            \n",
    "        input_frame = ctk.CTkFrame(self.master)\n",
    "        input_frame.pack(pady=10)\n",
    "        # Name label\n",
    "        name_label = ctk.CTkLabel(input_frame, text=\"Pattern \")\n",
    "        name_label.grid(row=0, column=0, padx=5, pady=5, sticky=\"e\")\n",
    "        # Entry box for name(pattern)\n",
    "        name_entry = ctk.CTkEntry(input_frame)\n",
    "        name_entry.grid(row=0, column=1, padx=5, pady=5)\n",
    "\n",
    "        # Frame for results\n",
    "        result_frame = ctk.CTkFrame(self.master)\n",
    "        result_frame.pack(pady=10)\n",
    "\n",
    "        result_text = tk.Text(result_frame, width=80, height=20, state=tk.DISABLED)\n",
    "        result_text.pack()\n",
    "\n",
    "        def run_algo():\n",
    "            nonlocal selected_algo, result_text\n",
    "            result_text.config(state=tk.NORMAL)\n",
    "            result_text.delete(1.0, tk.END)\n",
    "            \n",
    "            \n",
    "\n",
    "            if not self.dna_sequence:\n",
    "                result_text.insert(tk.END, \"Please select a DNA file first.\\n\")\n",
    "                result_text.config(state=tk.DISABLED)\n",
    "                return\n",
    "            pattern=name_entry.get()\n",
    "            algo = selected_algo.get()\n",
    "            result = self.run_algorithm(algo, self.dna_sequence,pattern)\n",
    "            # Insert the results into the text widget\n",
    "            result_text.insert(tk.END, f\"Results for {algo}:\\n\", \"header\")\n",
    "            if algo==algos[0]:\n",
    "                result_text.insert(tk.END, f\"\\n DNA reverse complementary: \\n\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result, \"DNA reverse complementary\")#\"\" this return algo\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.config(state=tk.DISABLED)\n",
    "            if algo==algos[1]:\n",
    "                result_text.insert(tk.END, f\"\\n Full translation:\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[0], \"\\n Full translation \\n\")\n",
    "                result_text.insert(tk.END, f\"\\n Protein translation:\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[1], \"\\n Protein translation \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.config(state=tk.DISABLED)\n",
    "            \n",
    "\n",
    "            if algo==algos[2]:\n",
    "                result_text.insert(tk.END, f\"\\n Pattern that we need:\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[0], \"\\n Pattern that we need \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                \n",
    "                \n",
    "                result_text.insert(tk.END, f\"\\n Pattern found at positions:\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[1] , \"\\n Start positions: \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.insert(tk.END, result[2] , \"\\n End positions: \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                \n",
    "                \n",
    "                \n",
    "                \n",
    "                result_text.insert(tk.END, f\"\\n the pattern that founded:\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[3], \"\\n the pattern that founded: \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.config(state=tk.DISABLED)\n",
    "                \n",
    "            if algo==algos[3]:\n",
    "                result_text.insert(tk.END, f\"\\n Position of matched pattern :\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[0], \"\\n position \\n\") \n",
    "                result_text.insert(tk.END, f\"\\n Skip Alignment in Bad Character rule :\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[1], \"\\n Bad \\n\") \n",
    "                result_text.insert(tk.END, f\"\\n Skip Alignment in Good Suffix rule  :\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[2], \"\\n Good \\n\") \n",
    "                result_text.config(state=tk.DISABLED)\n",
    "            if algo==algos[4]:\n",
    "                result_text.insert(tk.END, f\"\\n Position of matched pattern :\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[0], \"\\n position \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.insert(tk.END, f\"\\n Skipped :\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[1], \"\\n Skiiped \\n\") \n",
    "                result_text.config(state=tk.DISABLED)\n",
    "            \n",
    "            if algo==algos[5]:\n",
    "                result_text.insert(tk.END, \"\\n Occurrences of pattern:\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result, \"\\n Occurrences \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.config(state=tk.DISABLED)\n",
    "            \n",
    "            if algo==algos[-2]:\n",
    "                result_text.insert(tk.END, f\"\\n Position Match:\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[0], \"\\n Position \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.insert(tk.END, f\"\\n Skipped:\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[1], \"\\n Skipped \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.config(state=tk.DISABLED)\n",
    "    \n",
    "            if algo==algos[-1]:\n",
    "                result_text.insert(tk.END, f\"\\n the approximately matching position that founded:\\n\", \"header\")\n",
    "                result_text.insert(tk.END, result[0], \"\\n the Start Position: \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.insert(tk.END, result[1] , \"\\n End positions: \\n\")\n",
    "                result_text.insert(tk.END, \"\\n\")\n",
    "                result_text.config(state=tk.DISABLED)\n",
    "                \n",
    "            #if algo==algos[4]:\n",
    "            #if algo==algos[5]:\n",
    "            #if algo==algos[6]:\n",
    "            #if algo==algos[7]:\n",
    "                    \n",
    "        run_button = ctk.CTkButton(self.master, text=\"Run algorithm\", command=run_algo)\n",
    "        run_button.pack(pady=10)\n",
    "\n",
    "        # Configure text widget tags\n",
    "        result_text.tag_configure(\"header\", font=(\"Arial\", 15, \"bold\"), underline=True,foreground=\"red\")\n",
    "        \n",
    "        \n",
    "        \n",
    "        result_text.tag_configure(\"DNA reverse complementary\", font=(\"Arial\", 12,))\n",
    "        result_text.tag_configure(\"Full translation\", font=(\"Arial\", 12,\"italic\"))\n",
    "        result_text.tag_configure(\"Protein translation\", font=(\"Arial\", 12, \"italic\"))\n",
    "\n",
    "    def run_algorithm(self, algo, sequence,pattern):\n",
    "        if algo == \"DNA reverse complementary\":\n",
    "            return self.reverse_complement(sequence)\n",
    "        elif algo == \"DNA translation to protein\":\n",
    "            return self.translate(sequence)\n",
    "        elif algo == \"Exact matching\":\n",
    "            return self.naive_search(sequence,pattern)\n",
    "        elif algo == \"Boyer moore\":\n",
    "            return self.boyer_moore(sequence,pattern)\n",
    "        elif algo == \"K-mer indexing\":\n",
    "            return self.kmer_indexing(sequence,pattern)\n",
    "        elif algo == \"K-M-P algorithm\":\n",
    "            return self.kmp(sequence,pattern)\n",
    "        elif algo == \"Suffix array\":\n",
    "            return self.suffix_array(sequence,pattern)\n",
    "        elif algo == \"Approximate matching\":\n",
    "            return self.approximate_matching(sequence,pattern)\n",
    "\n",
    "        \n",
    "# implementation of all algorithms\n",
    "    def reverse_complement(self, sequence):\n",
    "        complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "        reverse_complement = ''.join(complement.get(base, base) for base in reversed(sequence))\n",
    "        return reverse_complement\n",
    "\n",
    "    def translate(self, sequence):\n",
    "        dic = {\"TTT\" : \"F\", \"CTT\" : \"L\", \"ATT\" : \"I\", \"GTT\" : \"V\",\n",
    "               \"TTC\" : \"F\", \"CTC\" : \"L\", \"ATC\" : \"I\", \"GTC\" : \"V\",\n",
    "               \"TTA\" : \"L\", \"CTA\" : \"L\", \"ATA\" : \"I\", \"GTA\" : \"V\",\n",
    "               \"TTG\" : \"L\", \"CTG\" : \"L\", \"ATG\" : \"M\", \"GTG\" : \"V\",\n",
    "               \"TCT\" : \"S\", \"CCT\" : \"P\", \"ACT\" : \"T\", \"GCT\" : \"A\",\n",
    "               \"TCC\" : \"S\", \"CCC\" : \"P\", \"ACC\" : \"T\", \"GCC\" : \"A\",\n",
    "               \"TCA\" : \"S\", \"CCA\" : \"P\", \"ACA\" : \"T\", \"GCA\" : \"A\",\n",
    "               \"TCG\" : \"S\", \"CCG\" : \"P\", \"ACG\" : \"T\", \"GCG\" : \"A\",\n",
    "               \"TAT\" : \"Y\", \"CAT\" : \"H\", \"AAT\" : \"N\", \"GAT\" : \"D\",\n",
    "               \"TAC\" : \"Y\", \"CAC\" : \"H\", \"AAC\" : \"N\", \"GAC\" : \"D\",\n",
    "               \"TAA\" : \"*\", \"CAA\" : \"Q\", \"AAA\" : \"K\", \"GAA\" : \"E\",\n",
    "               \"TAG\" : \"*\", \"CAG\" : \"Q\", \"AAG\" : \"K\", \"GAG\" : \"E\",\n",
    "               \"TGT\" : \"C\", \"CGT\" : \"R\", \"AGT\" : \"S\", \"GGT\" : \"G\",\n",
    "               \"TGC\" : \"C\", \"CGC\" : \"R\", \"AGC\" : \"S\", \"GGC\" : \"G\",\n",
    "               \"TGA\" : \"*\", \"CGA\" : \"R\", \"AGA\" : \"R\", \"GGA\" : \"G\",\n",
    "               \"TGG\" : \"W\", \"CGG\" : \"R\", \"AGG\" : \"R\", \"GGG\" : \"G\" \n",
    "               }\n",
    "        s=\"\"\n",
    "        sf=\"\"\n",
    "        flag=0\n",
    "        for i in range(0,len(sequence)-2,3):\n",
    "            if dic[sequence[i:i+3]]==\"M\":\n",
    "                flag=1\n",
    "            elif (dic[sequence[i:i+3]]==\"*\"):\n",
    "                flag=0\n",
    "\n",
    "            if flag==1:\n",
    "                s+=dic[sequence[i:i+3]]\n",
    "            sf+=dic[sequence[i:i+3]]\n",
    "        # Return both the full translation sequence and the protein sequence as a tuple\n",
    "        return  sf,s\n",
    "\n",
    "    def naive_search(self, sequence,pattern):\n",
    "        matches=[]\n",
    "        pattern_len=len(pattern)\n",
    "        sequence_len=len(sequence)\n",
    "        for i in range(sequence_len-pattern_len+1):\n",
    "            if sequence[i:i+pattern_len]==pattern:\n",
    "                matches.append(i)\n",
    "        if not matches:\n",
    "            return f'Pattern that we need to search {pattern}',f'Start Position not founded',f'End Position not founded',f'Sequence that you need not in the dna sequence'\n",
    "        return f'Pattern that we need to search {pattern}',f'Start Position {matches[0]}',f'End Position {matches[0]+len(pattern)}',f'Sequence that we founded it {sequence[matches[0]:matches[0]+len(pattern)]}'\n",
    "\n",
    "    def approximate_matching(self, sequence,pattern):\n",
    "        dp= [[0 for i in range(len(pattern)+1)] for j in range(len(sequence)+1)]\n",
    "\n",
    "        for j in range(len(pattern)+1):\n",
    "            dp[0][j]=j\n",
    "        for i in range(1,len(sequence)+1):\n",
    "            for j in range(1,len(pattern)+1):\n",
    "                if sequence[i-1]==pattern[j-1]:\n",
    "                    dp[i][j]=dp[i-1][j-1]\n",
    "                else:\n",
    "                    dp[i][j]=min(dp[i-1][j-1],dp[i-1][j],dp[i][j-1])+1\n",
    "                    \n",
    "        matches=[]\n",
    "        for i in range(1,len(sequence)+1):\n",
    "            if dp[i][len(pattern)] == 0:\n",
    "                matches.append(i-len(pattern))\n",
    "        if not matches:\n",
    "            return f'Start Position not founded',f'End Position not founded'\n",
    "        return f'Start position : {matches[0]}', f'End Position : {matches[0]+len(pattern)}'\n",
    "    \n",
    "\n",
    "    def boyer_moore(self,sequence, pattern):\n",
    "        skip_bad = {}\n",
    "        for i in range(len(pattern)-1):\n",
    "            skip_bad[pattern[i]] = len(pattern)-i-1\n",
    "        # Adding Good Suffix rule\n",
    "        suffixes = _build_suffix_array(pattern)\n",
    "        skip_good = {}\n",
    "        for i in range(len(pattern)-1):\n",
    "            j = suffixes[i]\n",
    "            skip_val = len(pattern)-j-1\n",
    "            if skip_val not in skip_good:\n",
    "                skip_good[len(pattern)-j-1] = skip_val\n",
    "        # Applying Bad Character rule\n",
    "        i = 0\n",
    "        while i <= len(sequence)-len(pattern):\n",
    "            j = len(pattern)-1\n",
    "            while j >= 0 and pattern[j] == sequence[i+j]:\n",
    "                j -= 1\n",
    "            if j < 0:\n",
    "                return i-1, len(skip_bad), len(skip_good)\n",
    "            else:\n",
    "                skip_val = skip_bad.get(sequence[i+j], len(pattern))\n",
    "                i += skip_val\n",
    "        return -1, len(skip_bad), len(skip_good), i\n",
    "    \n",
    "    def suffix_array(self,sequence,pattern):\n",
    "        suffix_array = _build_suffix_array(sequence)\n",
    "        matches = []\n",
    "        left = 0\n",
    "        right = len(sequence)\n",
    "        alignments = 0\n",
    "        while left < right:\n",
    "            middle = (left + right) // 2\n",
    "            suffix = suffix_array[middle]\n",
    "            alignments += 1\n",
    "            if pattern > sequence[suffix:suffix + len(pattern)]:\n",
    "                left = middle + 1\n",
    "            else:\n",
    "                right = middle\n",
    "        start = left\n",
    "        right = len(sequence)\n",
    "        while left < right:\n",
    "            middle = (left + right) // 2\n",
    "            suffix = suffix_array[middle]\n",
    "            alignments += 1\n",
    "            if pattern < sequence[suffix:suffix + len(pattern)]:\n",
    "                right = middle\n",
    "            else:\n",
    "                left = middle + 1\n",
    "        end = right\n",
    "        for i in range(start, end):\n",
    "            suffix = suffix_array[i]\n",
    "            matches.append(suffix)\n",
    "        skipped = (len(sequence) - len(pattern) + 1) - alignments\n",
    "        matches.sort()\n",
    "        return matches,skipped\n",
    "\n",
    "    def _build_suffix_array(self,sequence):\n",
    "        suffixes = [(sequence[i:], i) for i in range(len(sequence))]\n",
    "        suffixes.sort()\n",
    "        return [suffix[1] for suffix in suffixes]\n",
    "\n",
    "    def kmer_indexing(self, sequence, pattern):\n",
    "        \n",
    "        kmer_dict = self._build_kmer_index(sequence,min(len(pattern),50))\n",
    "        matches = []\n",
    "        kmer = pattern[:min(len(pattern),50)]\n",
    "        alignments = 0\n",
    "        if kmer in kmer_dict:\n",
    "            for index in kmer_dict[kmer]:\n",
    "                alignments += 1\n",
    "                if sequence[index:index+len(pattern)] == pattern:\n",
    "                    matches.append(index)\n",
    "        skipped = len(sequence) - len(pattern) + 1 - alignments\n",
    "        return matches, skipped\n",
    "    \n",
    "    def _build_kmer_index(self, sequence, k):\n",
    "        kmer_dict = {}\n",
    "        for i in range(len(sequence) - k + 1):\n",
    "            kmer = sequence[i:i+k]\n",
    "            if kmer not in kmer_dict:\n",
    "                kmer_dict[kmer] = []\n",
    "            kmer_dict[kmer].append(i)\n",
    "        return kmer_dict\n",
    "    def kmp(self, sequence,pattern):\n",
    "        failure_table = self.create_failure_table(pattern)\n",
    "        i = 0\n",
    "        j = 0\n",
    "        occurrences = []\n",
    "        while i < len(sequence):\n",
    "            if pattern[j] == sequence[i]:\n",
    "                i += 1\n",
    "                j += 1\n",
    "                if j == len(pattern):\n",
    "                    occurrences.append(i - len(pattern))\n",
    "                    j = failure_table[j - 1]\n",
    "            else:\n",
    "                if j > 0:\n",
    "                    j = failure_table[j - 1]\n",
    "                else:\n",
    "                    i += 1\n",
    "        return f\"Occurrences of '{pattern}': {occurrences}\"\n",
    "    def create_failure_table(self, pattern):\n",
    "        failure_table = [0] * len(pattern)\n",
    "        i = 1\n",
    "        j = 0\n",
    "        while i < len(pattern):\n",
    "            if pattern[i] == pattern[j]:\n",
    "                j += 1\n",
    "                failure_table[i] = j\n",
    "                i += 1\n",
    "            else:\n",
    "                if j != 0:\n",
    "                    j = failure_table[j - 1]\n",
    "                else:\n",
    "                    failure_table[i] = 0\n",
    "                    i += 1\n",
    "        return failure_table\n",
    "\n",
    "# Create the root window\n",
    "root = ctk.CTk()\n",
    "app = BioinformaticsApp(root)\n",
    "root.mainloop()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdcc2d89",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
